logstash-input-tcp.gemspec

```diff
@@ -23,10 +23,16 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency "logstash-core-plugin-api", ">= 1.60", "<= 2.99"
   s.add_runtime_dependency 'logstash-mixin-ecs_compatibility_support', '~>1.2'
 
-  s.add_runtime_dependency 'logstash-core', '>= 6.7.0'
-
-  # we depend on bouncycastle's bcpkix-jdk15on being on the class-path
-  s.add_runtime_dependency 'jruby-openssl', '>= 0.10.2'
+  s.add_runtime_dependency 'logstash-core', '>= 6.7.0' # LS 6.7: JRuby 9.2.7.0
+
+  # NOTE: be able to test against LS 7.x / 8.x
+  # (some LS versions have strict jruby-openssl requirements in logstash-core e.g. "= 0.11.0")
+  if ENV['ELASTIC_CONTAINER'].eql?('true') && ENV['PWD'].eql?('/usr/share/plugins/plugin')
```

Comment:
> So under CI we simply don't declare a runtime dependency on JOSSL?

correct we assume the one from logstash-core to be used (which is either smt limited e.g. `= 0.11.0` or latest due smt like `~> 0.11.0`).

> How do we ensure that we don't build a gem artifact that is missing this dependency declaration?

we assume the gem is not to be build in a docker container used for testing (plugin at `/usr/share/plugins/plugin`). I am open to better suggestions - this is an ugly compromise until we unlock jossl version in LS 7.x / 8.x (pending)

Reply:
we can not provide all the features we ship here wout the latest JOSSL, tests simply work-around and not test the features when up-to-date JOSSL isn't available (this is to change later when old LS versions aren't locked to a specific JOSSL version - e.g. we recently unlocked 7.17 to not limit JOSSL to `= 0.11.1`).

- in client mode TLSv1.3 is only supported since JOSSL 0.12
- the ability to set Java cipher names on the Ruby SSLContext (client mode) only works since 0.12.2
- installing latest tcp plugin version is expected to work on 7.x (when jossl isn't strictly locked) as well as 8.x moving forward (in 8.x again the strict lock has been relaxed back)

Reply:
the JOSSL features are mentioned through the code as comments:
- https://github.com/logstash-plugins/logstash-input-tcp/pull/198/files#diff-012c902c5ac82e3d122ada84ad7e711f565ff0187dabc4cca2e3a56215ba05b3R313
- https://github.com/logstash-plugins/logstash-input-tcp/pull/198/files#diff-48a0ca35b36296ce60b56aadd06fdbee9dfd628a74e0a69138157d430ca78cb1R34

Reply:
> we can not provide all the features we ship here wout the latest JOSSL,

This makes sense.

> tests simply work-around and not test the features when up-to-date JOSSL isn't available 

If the distributable artifact of this plugin cannot be _installed_ on a Logstash whose JOSSL isn't latest (because of pinning in Logstash core), why are we validating a subset of this plugin's behaviour in the presence of a potentially-older JOSSL at all? Should we instead gatekeep the _feature_ and warn/error when it is invoked without the prerequisite met?

***

If a subset of the functionality introduced in this PR is not available on older Logstashes (that have JOSSL pinned), and we include a dependency declaration in our gem artifact that prevents them from being installed on those older Logstashes, why are we going out of our way in CI to ensure that the rest of the functionality works? None of the functionality will work because the plugin manager will refuse to install it.

Reply:
so your proposal is to e.g. have a :red_circle: 7.x built (at least until 7.17.2 is released which unlocks jossl) ?
when this feature started 8.x has locked jossl as well.

the build is going to fail in the image build step (during `bundle install`)

Reply:
It is pointless to have green CI for a version of Logstash upon which the distributable version of the artifact cannot be installed. 

If some subset of the functionality is expected to work with older JOSSL, then newest JOSSL isn't actually a runtime requirement of the whole.

My proposal is that we specify the minimum JOSSL we _actually_ require, and reject configuration _at runtime_ when that configuration requires a newer JOSSL than we have. For example, if specifying ciphers requires a newer JOSSL than is available at runtime, we raise a config error if and only if the user specifies ciphers. Similarly, if TLSv1.3 requires a newer JOSSL than we have, we raise a config error if and only if the user specifies TLSv1.3.

Reply:
> It is pointless to have green CI for a version of Logstash upon which the distributable version of the artifact cannot be installed.

as noted it was meant to be a "temporary measure" - to me it made sense to have a :green_circle: than for others to get confused why CI is failing for 7.x ... anyhow it's now :red_circle: as I do not care about my preference here if a :red_circle: CI is acceptable (as I forced the `>= 0.12.2` requirement).

going to look into setting up some older 7.x version manually but it does not feel like time well spent ...

Reply:
added extra runtime check on Java 8 for when `TLSv1.3` would not work I am not 100% it's worth the effort given Java 8 is effectively dead - the assumption I operated with was that documenting this is more than enough ... also other plugins, already shipped with the new `ssl_supported_protocols` setting, do not have this extra check atm.